### PR TITLE
Qmotif create writer thread that runs in background

### PR DESCRIPTION
**Pull request by holmeso**
# Description

During a recent piece of work where `qmotif` was run against a thousand bam files, it became apparent that it was running out of memory.
These runs were different to normal in that we were looking at all regions of the genome rather than just the starts and ends of chromosomes as is usually done.

What was happening was that the SAMRecords that contained the motifs of interest were being stored in a queue until all the processing was done, and then written to file. With much larger regions than normal, this meant that a large number of SAMRecords were stored in memory, resulting in the `OutOfMemoryErrors` and high thread use due to GC.

The solution was to create a thread that writes the SAMRecords to file at the start of the process. This writer thread picks up the SAMRecords as they are placed into the queue meaning that they don't sit in memory.

A couple of other minor fixes have also been included in this PR:

- cached the hashCode value in `ChrPositionName` - this was using a (relatively) large amount of cpu time as `ChrPositionName` is a key in a widely used map. Caching this reduces cpu use here.
 - Changed from having a `StringBuilder` to represent the motifs found in a region to a map. This is really weird as judging by the code, maps were used originally and then changed to StringBuilder. This is really not sensible as when running with large regions, there will be a huge number of (duplicated) entries in the `StringBuilder`. Reverted back.
 - fixed some styling on classes that were updated

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How Has This Been Tested?
Have run the updated code and compared against an existing output, and results were the same.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: qcommon/src/org/qcmg/common/model/ChrPositionName.java [hunk 1]
```diff
@@ -53,9 +54,14 @@ public String toIGVString() {
 	 */
 	@Override
 	public int hashCode() {
-		final int prime = 31;
-		int result = super.hashCode();
-		result = prime * result + ((name == null) ? 0 : name.hashCode());
+	    int result = hashCode;
+	    if (result == 0) {
+	    	result = 17;
+			final int prime = 31;
```
Comment (ChristinaXu2017) [inline qcommon/src/org/qcmg/common/model/ChrPositionName.java:60]: why bigger indent than last line?

File: qmotif/src/org/qcmg/motif/JobQueue.java [hunk 2]
```diff
@@ -146,35 +148,34 @@ public JobQueue(final Configuration invariants)
...
-		writeOutput();
-		logger.info("DONE - Writing output");
-		
 		logger.info("total number from input is " + countIn.get());
 		logger.info("total number of reads (satisfied by the query if query provided) count to coverage is " + countOut.get());
 		
 		logger.info("Performing final reduce step on results");
 		reduceResults();
 		logger.info("Final reduce step complete");
 //		logger.debug("Final reduced results: " + perIdPerCoverageBaseCounts);
```
Comment (ChristinaXu2017) [inline qmotif/src/org/qcmg/motif/JobQueue.java:173]: delete it

File: qmotif/test/org/qcmg/motif/util/RegionCounterTest.java [hunk 3]
```diff
@@ -45,17 +45,17 @@ public void excludedRegion() {
 	public void includedRegionUnmappedReadsFS() {
 		rcInc.addMotif("ABCD", true, false);
 		assertEquals(true, rcInc.hasMotifs());
-//		assertEquals(true, rcInc.getMotifsForwardStrand().containsKey("ABCD"));
-		assertEquals("ABCD", rcInc.getMotifsForwardStrand());
+		assertEquals(true, rcInc.getMotifsForwardStrand().containsKey("ABCD"));
+//		assertEquals("ABCD", rcInc.getMotifsForwardStrand());
```
Comment (ChristinaXu2017) [inline qmotif/test/org/qcmg/motif/util/RegionCounterTest.java:49]: should this commented line be removed?

Comment (holmeso) [reply qmotif/test/org/qcmg/motif/util/RegionCounterTest.java]: done

File: qmotif/src/org/qcmg/motif/JobQueue.java (hunk 2 above)
Comment (holmeso) [reply qmotif/src/org/qcmg/motif/JobQueue.java]: done

File: qcommon/src/org/qcmg/common/model/ChrPositionName.java (hunk 1 above)
Comment (holmeso) [reply qcommon/src/org/qcmg/common/model/ChrPositionName.java:60]: fixed

**Comment by holmeso** [comment]
thanks for your review Christina

**Pull request merged**